packages/lit-dev-cloudbuild-cleanup/src/cleanup.ts

```diff
@@ -0,0 +1,272 @@
+import {request} from 'gaxios';
+
+const ONE_WEEK_IN_MS = 7 * 24 * 60 * 60 * 1000;
+const ONE_WEEK_AGO = new Date(Date.now() - ONE_WEEK_IN_MS);
+
+const {PROJECT_ID, REPO_NAME, _DEPLOY_REGION} = process.env;
+
+if ([PROJECT_ID, REPO_NAME, _DEPLOY_REGION].some((env) => env === undefined)) {
+  console.log({PROJECT_ID, REPO_NAME, _DEPLOY_REGION});
+  throw new Error(
+    'missing one or more required environment variables: PROJECT_ID, REPO_NAME, _DEPLOY_REGION'
+  );
+}
+
+const sleep = (time: number) =>
+  new Promise((resolve) => setTimeout(resolve, time));
+
+async function main() {
+  const octokit = new Octokit();
+
+  const openPrs = new Set<number>();
+
+  console.log('fetching open PRs from github');
+  const openPrsIterator = octokit.paginate.iterator(octokit.rest.pulls.list, {
+    owner: 'lit',
+    repo: `${REPO_NAME}`,
+    state: 'open',
+    per_page: 100,
+  });
+
+  for await (const {data: prs} of openPrsIterator) {
+    for (const pr of prs) {
+      openPrs.add(pr.number);
+    }
+  }
+
+  console.log('found open prs', openPrs);
+
+  const run = google.run('v1');
+
+  const auth = new google.auth.GoogleAuth({
+    scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+  });
+
+  const authClient = await auth.getClient();
+  google.options({auth: authClient});
+
+  // fetch list of revisions across both services
+  console.log('fetching list of revisions');
+  const {data} = await run.projects.locations.revisions.list({
+    parent: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}`,
+  });
+
+  let revisions = data;
+
+  if (!revisions.items) {
+    throw new Error('found no revision items');
+  }
+
+  console.log('total revisions count', revisions.items.length);
+
+  const revisionsToKeepByRecency = new Set<string>();
+  for (const rev of revisions.items) {
+    if (!rev.metadata) {
+      throw new Error('found no metadata');
+    }
+
+    if (new Date(rev.metadata.creationTimestamp!) > ONE_WEEK_AGO) {
+      revisionsToKeepByRecency.add(rev.metadata.name!);
+    }
+  }
+
+  console.log(
+    'revisions younger than a week count',
+    revisionsToKeepByRecency.size
+  );
+
+  // traffic tags must be cleaned up before a revision can be deleted
+  const services = ['lit-dev', 'lit-dev-playground'];
+  let serviceUpdated = false;
+  await Promise.all(
+    services.map(async (serviceName) => {
+      console.log(`fetching ${serviceName} service`);
+      const {data} = await run.projects.locations.services.get({
+        name: `projects/${PROJECT_ID}/locations/${_DEPLOY_REGION}/services/${serviceName}`,
+      });
+
+      if (!data.spec || !data.spec.traffic) {
+        throw new Error(`found no traffic for ${serviceName} service`);
+      }
+
+      console.log(`${serviceName} traffic count`, data.spec.traffic.length);
+
+      const trafficToKeep = data.spec.traffic.filter((t) => {
+        return (
+          // precence of percent indicates traffic is being routed here (i.e. prod)
+          Boolean(t.percent) ||
```

Comment:
This is false for both `undefined` and `0`. Which seems reasonable, just checking that's the intention (maybe `0` means something subtly different?). Could consider checking both explicitly for clarity.

Reply:
The way we're using it, we should never see `0` as we always direct 100% of traffic to a single revision. https://cloud.google.com/run/docs/rollouts-rollbacks-traffic-migration#split-traffic
I don't believe cloud run even creates `0` values there unless the yaml is manually edited to be that way. And if the value somehow is `0`, we'll still keep it for a week based on the recency check. No traffic and older than a week seems safe to get rid of.